packages/shared/src/transactions.ts

```diff
@@ -154,15 +154,34 @@ export const isSentTx = (
 export const isWalletSelfTransfer = (
   tx: e.Transaction | e.PendingTransaction | e.MempoolTransaction,
   walletAddresses: AddressHash[]
-): boolean => getTxAddresses(tx).every((address) => address && walletAddresses.includes(getBaseAddressStr(address)))
+): boolean => getTxAddresses(tx).every((address) => walletAddresses.includes(getBaseAddressStr(address)))
 
 export const isContractTx = (tx: e.Transaction | e.PendingTransaction | e.MempoolTransaction): boolean =>
   !!tx.outputs?.some(inputOutputIsContractAddress) || !!tx.inputs?.some(inputOutputIsContractAddress)
 
 const inputOutputIsContractAddress = (io: e.Input | e.Output): boolean => !!io.address && isContractAddress(io.address)
 
-const getTxAddresses = (tx: e.Transaction | e.PendingTransaction | e.MempoolTransaction): (AddressHash | undefined)[] =>
-  [...(tx.outputs ?? []), ...(tx.inputs ?? [])].map((io) => io.address)
+const getTxAddresses = (tx: e.Transaction | e.PendingTransaction | e.MempoolTransaction): AddressHash[] => {
+  const addresses = new Set<AddressHash>()
+
+  if (tx.outputs) {
+    for (const output of tx.outputs) {
+      if (output.address && !addresses.has(output.address)) {
```

Comment:
A `Set` ensures unique values by default, so I don't think this check is needed. If I'm not mistaken, I think this could be simplify quite a bit, and only use a single loop:

```ts
const getTxAddresses = (
  tx: e.Transaction | e.PendingTransaction | e.MempoolTransaction
): AddressHash[] => {
  const addresses = new Set<AddressHash>()
  const entries = [...(tx.outputs ?? []), ...(tx.inputs ?? [])]

  for (const entry of entries) {
    if (entry.address) {
      addresses.add(entry.address)
    }
  }

  return Array.from(addresses)
}

```

Reply:
You are right, the check is not needed.

I want to avoid deconstructing and creating new objects unnecessarily. But I measured the difference and even though your version is slightly worse I think it's acceptable cause the difference is small. Updating...

Reply:
Pushed